tokio-stream/src/lib.rs

```diff
@@ -81,6 +81,15 @@ pub mod wrappers;
 
 mod stream_ext;
 pub use stream_ext::{collect::FromStream, StreamExt};
+/// Adapters for types that implement the `Stream` trait.
+/// Created through the tokio [`StreamExt`] methods.
+pub mod adapters {
+    pub use crate::stream_ext::{
+        AllFuture, AnyFuture, Chain, Filter, FilterMap, FoldFuture, Fuse, Map, MapWhile, Merge,
+        Next, Peekable, Skip, SkipWhile, Take, TakeWhile, Then, TryNext,
+    };
+}
+
 cfg_time! {
     pub use stream_ext::timeout::{Elapsed, Timeout};
```

Comment:
And I just saw that PR #4601 exposed `Timeout` directly in `tokio_stream`. What to do about this ?

Expose them all in `adapters`, but also keep this one exposed flush in `tokio_stream` to not break anything?

Reply:
You can make the top-level ones `#[doc(hidden)]`.

Reply:
Done, and added the error struct Elapsed in the new imports, so that's it's still somewhere to be found if a user who used it wants to update.

Reply:
Looks like the tests don't like the addition of `#[doc(hidden)]`...

Reply:
@obi1kenobi is there any way to tell cargo-semver-checks to treat a `#[doc(hidden)]` item as part of the public API? Context is that we want to change the path of an exported item.

Reply:
Assuming the hidden item you're asking about is this re-export of `tokio_stream::stream_ext::timeout::{Elapsed, Timeout};`, I'd recommend marking the re-export as `#[deprecated]`.

Deprecated items are [always public API regardless of `#[doc(hidden)]`](https://predr.ag/blog/checking-semver-for-doc-hidden-items/#what-about-deprecated-items).

I believe this:
- won't cause the underlying item to be considered deprecated
- will cause this re-export to be considered deprecated
- won't cause a `cargo-semver-checks` lint for removal from public API.

If you don't think this is a good solution for any reason, I'd love to chat more about it! It would be useful user research to guide future `cargo-semver-checks` development.

Reply:
I have tried using `#[deprecated]` locally. I does remove the error thrown by `cargo-semver-checks`, but
- no deprecation notice appears in the generated doc (I would expect it to appear somewhere on [this page](https://docs.rs/tokio-stream/latest/tokio_stream/struct.Timeout.html))
- I see no warning when compiling a test in which I put `use tokio_stream::Timeout` and used `Timeout`

The docs on the `deprecated` attribute [do say](https://doc.rust-lang.org/nightly/reference/attributes/diagnostics.html#the-deprecated-attribute) that 'any item' can be marked as such. Am I looking at the wrong place? What does it do for a re-export to be considered deprecated as you say?

Reply:
Actually, it seems that we can't deplicate re-export yet :(

https://github.com/rust-lang/rust/issues/30827

Reply:
Marking the _re-export_ as deprecated doesn't have any effect in rustc / clippy / generated rustdoc HTML (it is simply ignored) but does have the desirable effect in `cargo-semver-checks`.

Unless I've missed something, to me that still sounds fine to use. After all, it didn't seem like originally (before I suggested it) you were planning to deprecate the re-export, so you weren't going to get any rustc / clippy / HTML docs notices on it anyway.

I don't think anything bad happens here when rustc / clippy / generated docs start fully supporting deprecations on re-exports.

> What does it do for a re-export to be considered deprecated as you say?

Deprecations and `doc(hidden)`-ness are properties of import paths like `tokio_stream::stream_ext::timeout::Timeout`, not the items behind those names. The full details are [in this post](https://predr.ag/blog/checking-semver-for-doc-hidden-items/), but as a summary:
```rust
#[doc(hidden)]
pub mod example {
    pub struct Something;
}

pub use example::Something;
```
In this example, `this_crate::Something` is public API because in writing that path, we didn't touch any hidden items along the way. `this_crate::example::Something` is hidden and non-public API because the `this_crate::example` prefix is hidden. Conceptually, you can imagine walking the components `["this_crate", "example", "Something"]` and checking whether each of them is hidden or deprecated — if so, the entire path is hidden / deprecated as well.

Broadly speaking, the rustc / clippy / rustdoc HTML implementations only consider deprecations as properties of items like `pub struct Something`, not of import paths. (There are some subtleties — check out the linked post for more details.) `cargo-semver-checks` used to do the same thing — it's much simpler and it isn't a priori obvious that deprecation is a path property. This led to tons of false-positives so I spent a couple of months to fix it, as the post describes.

Reply:
Thank you for the in-depth message. Now I have the full picture, and - as you say on your blog - the semantics are perfectly aligned with what we want.

Reply:
Awesome! As always, please feel free to ping me if you have further questions or run into any issues :+1:

Reply:
Thank you @obi1kenobi. I agree that using `#[deprecated]` is reasonable in this case.